RFC-0017-PyTorch-Operator-Versioning.md

```diff
@@ -0,0 +1,223 @@
+        * This seems straightforward for Torchscript and Edge testing but I’m not sure how it would work for deploy/package
+    * [Improved FC testing] Tests that the new version of the operator can still be loaded on old runtimes and run as expected need to be easy to add 
+        * This might require a new test job, which could be tricky to setup
+* **Torchscript changes**
+    * TorchScript Runtime build contains a table of operators and their corresponding versions
+    * Each serialized model contains a table of operators and their corresponding version according to the TorchScript compiler that generated the model
+        * Q: will this require a new serialization format?
+        * Yes, but it is easy to make it BC and FC
+    * During loading into the TorchScript compiler, TorchScript needs to match operator schema according to the table of operator versions stored in the package. This would generate IR that conforms to older schema. 
+    * Then a TorchScript pass takes IR of older schema and replaces older versions of operator invocations with bodies of upgraders.
+    * Out-of-support operator versions (ie. those no longer defined in `native_functions.yaml` with a valid upgrader) need to throw an error
+* **Edge runtime and mobile delivery service changes**
+    * Delivery compatibility: communicating operator version table deployed on device and deliver models appropriately
+    * Runtime: load upgraders at model loading time, to ensure older models always work after updating runtime
+    * Unknown operator versions need to throw an error
+    * The operator version and upgraders are built into the runtime for BC.
+    * Allow for the addition of optional keyword-only arguments without a version bump or FC concern
+    * Since additional operators can be introduced in upgraders, tracing based selective build should also cover upgraders: easier for BC because the new runtimes goes with the upgraders. 
+* **torch.package changes**
+    * Each torch.package package contains a table of operators and corresponding version according to PyTorch build used to package the model
+        * Q: How does the torch.package scenario for mapping old versions to current PyTorch operators work?
+        * A: Operator versioning, by design, can’t cover all torch.package use cases. So this should be out of scope.
+* **New documentation required**
+    * e2e BC-breaking guide
+        * To make a BC-breaking change update the version and write a torchscript adaptor and a mobile adaptor
+    * e2e FC-breaking guide
+        * It’s OK to add new optional keyword-only arguments as long as their default semantic preserve the operator’s current semantics
+
+Note that the proposal does not introduce an explicit version to _all_ PyTorch operators. Instead code changes are only required for updated operators with BC/FC breakage, that cannot be handled by automatic BC/FC methods. For other operators, the implicit version is v0.
```

Comment:
I'm not sure what code changes would be done for FC-breaking changes?

Reply:
No code changes required for FC-breaking, except the version bump.

Reply:
Currently we don't bump versions on FC-breaking changes. What's the proposal for bumping and what would bumping accomplish?